### ISSUE TITLE
proxyconnect tcp: dial tcp :0: getsockopt: connection refused when daemon running as ubuntu service

### ISSUE THREAD

**Issue by vladcalin** [moby/moby#35226]
**Description**

I have a weird issue with the docker daemon when running on ubuntu server after I installed it with ``sudo apt-get install docker-ce``. When I try to run any docker command that needs to go through the daemon, such as ``docker login`` or ``docker image pull redis`` I get the following error: ``Error response from daemon: Get https://registry-1.docker.io/v2/: proxyconnect tcp: dial tcp :0: getsockopt: connection refused``.

The weird part is that when I stop the docker service and start the daemon manually with ``sudo dockerd`` everything works as expected. 

Also, everything worked fine until a few days ago.

**Steps to reproduce the issue:**

I did not mange to reproduce the issue on another Ubuntu server. I am hoping you can provide some extra steps that will help trace the issue.

All I do is ``sudo docker pull redis`` and I get 
```
Using default tag: latest
Error response from daemon: Get https://registry-1.docker.io/v2/: proxyconnect tcp: dial tcp :0: getsockopt: connection refused
```
Only when the running the docker daemon as a service.

**Describe the results you received:**

```
Using default tag: latest
Error response from daemon: Get https://registry-1.docker.io/v2/: proxyconnect tcp: dial tcp :0: getsockopt: connection refused
```

**Describe the results you expected:**

```
Using default tag: latest
latest: Pulling from library/redis
Digest: sha256:07e7b6cb753f8d06a894e22af30f94e04844461ab6cb002c688841873e5e5116
Status: Image is up to date for redis:latest
```

**Additional information you deem important (e.g. issue happens only occasionally):**

Issue happen only when running the ``dockerd`` as a service. It does not happend when I start the daemon from terminal.

**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:18 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:56 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 14
Server Version: 17.09.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.10.0-37-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.794GiB
Name: Ubuntu-1604-xenial-64-minimal
ID: WQJY:HV5M:ZOKT:CPEN:ADCF:M4O2:UU7X:4M6J:WWS2:2P3O:QEHG:TSRQ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 18
 Goroutines: 25
 System Time: 2017-10-17T10:36:54.747218729+02:00
 EventsListeners: 0
Http Proxy:  HTTPS_PROXY= NO_PROXY=
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

The dockerd debug lines when the issue occurs: 
```
Oct 17 10:27:31 Ubuntu-1604-xenial-64-minimal dockerd[13445]: time="2017-10-17T10:27:31.592552414+02:00" level=debug msg="Calling GET /_ping"
Oct 17 10:27:31 Ubuntu-1604-xenial-64-minimal dockerd[13445]: time="2017-10-17T10:27:31.593528608+02:00" level=debug msg="Calling GET /v1.32/info"
Oct 17 10:27:31 Ubuntu-1604-xenial-64-minimal dockerd[13445]: time="2017-10-17T10:27:31.627050830+02:00" level=debug msg="Calling POST /v1.32/images/create?fromImage=redis&tag=latest"
Oct 17 10:27:31 Ubuntu-1604-xenial-64-minimal dockerd[13445]: time="2017-10-17T10:27:31.627264982+02:00" level=debug msg="Trying to pull redis from https://registry-1.docker.io v2"
Oct 17 10:27:31 Ubuntu-1604-xenial-64-minimal dockerd[13445]: time="2017-10-17T10:27:31.627706043+02:00" level=warning msg="Error getting v2 registry: Get https://registry-1.docker.io/v2/: proxyconnect tcp: dial tcp :0: getsockopt: connection refused"
Oct 17 10:27:31 Ubuntu-1604-xenial-64-minimal dockerd[13445]: time="2017-10-17T10:27:31.627761249+02:00" level=info msg="Attempting next endpoint for pull after error: Get https://registry-1.docker.io/v2/: proxyconnect tcp: dial tcp :0: getsockopt: connection refused"
Oct 17 10:27:31 Ubuntu-1604-xenial-64-minimal dockerd[13445]: time="2017-10-17T10:27:31.627787529+02:00" level=error msg="Handler for POST /v1.32/images/create returned error: Get https://registry-1.docker.io/v2/: proxyconnect tcp: dial tcp :0: getsockopt: connection refused"
```

**Comment by thaJeztah** [comment]
do you have a proxy configured in the systemd unit-file? I think `systemctl cat docker.service` will print all settings defined there

**Comment by thaJeztah** [comment]
Actually; looks like there's something configured indeed, looking at he output of your `docker info`; as this is a configuration issue on your system, not a bug, I'll close this issue for housekeeping, but feel free to continue the conversation if you need more help

**Comment by vladcalin** [comment]
The output of ``systemctl cat docker.service`` is

```
# /lib/systemd/system/docker.service
[Unit]
Description=Docker Application Container Engine
Documentation=https://docs.docker.com
After=network-online.target docker.socket firewalld.service
Wants=network-online.target
Requires=docker.socket

[Service]
Type=notify
# the default is not to use systemd for cgroups because the delegate issues still
# exists and systemd currently does not support the cgroup feature set required
# for containers run by docker
ExecStart=/usr/bin/dockerd -H fd://
ExecReload=/bin/kill -s HUP $MAINPID
LimitNOFILE=1048576
# Having non-zero Limit*s causes performance problems due to accounting overhead
# in the kernel. We recommend using cgroups to do container-local accounting.
LimitNPROC=infinity
LimitCORE=infinity
# Uncomment TasksMax if your systemd version supports it.
# Only systemd 226 and above support this version.
TasksMax=infinity
TimeoutStartSec=0
# set delegate yes so that systemd does not reset the cgroups of docker containers
Delegate=yes
# kill only the docker process, not all processes in the cgroup
KillMode=process
# restart the docker process if it exits prematurely
Restart=on-failure
StartLimitBurst=3
StartLimitInterval=60s

[Install]
WantedBy=multi-user.target

# /etc/systemd/system/docker.service.d/override.conf
[Service]
ExecStart=
ExecStart=/usr/bin/dockerd
# /etc/systemd/system/docker.service.d/proxy.conf
[Service]
Environment="HTTP_PROXY= HTTPS_PROXY= NO_PROXY="
```

There is no proxy configured. I can't see anything that might interfere with the daemon. It works when I start the daemon from the terminal with the ``sudo dockerd`` command.

**Comment by thaJeztah** [comment]
Looking at this file, which sets proxy environment vars, and is not part of the default package;

```
# /etc/systemd/system/docker.service.d/proxy.conf
[Service]
Environment="HTTP_PROXY= HTTPS_PROXY= NO_PROXY="
```

Can you try removing that file, reloading the systemd configuration and restarting docker?

**Comment by vladcalin** [comment]
Unfortunately I had to destroy the machine on which the issue was reproducing because I couldn't afford to keep it in that non-functional state any longer.

As a note, I used the https://github.com/angstwad/docker.ubuntu ansible role to install docker on that machine. I just saw that in the past days some proxy related commits were merged into that role, around the time when this issue started to appear. Indeed it seems not to be an docker related issue, but more of a configuration specific one.

Thank you for you time @thaJeztah !
